### PR TITLE
Luxfloat Optimisation and new defaults

### PR DESCRIPTION
**Pull request by borisbstyle**
Luxfloat PID controller was a nice concept, but very hard to tune and not as locked in as some others. After a lot of playing with different PID controllers I figured out what is good and what not.
Well this is how it should be! delta calculation was not optimal.
Please give some feedback

Updated:
- New defaults
- New delta calculation (RateError instead of GyroError)
- Removed TPA for Integral (not needed)

Flight with this PR and RC smoothing:
https://www.youtube.com/watch?v=1_SiBKD7R7M

NAZE
https://dl.dropboxusercontent.com/u/31537757/cleanflight/cleanflight_NAZE.hex

SRF3
https://dl.dropboxusercontent.com/u/31537757/cleanflight/cleanflight_SPRACINGF3.hex

CC3D:
https://dl.dropboxusercontent.com/u/31537757/cleanflight/cleanflight_CC3D.hex
https://dl.dropboxusercontent.com/u/31537757/cleanflight/cleanflight_CC3D.bin

### CODE REVIEW

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -192,16 +192,17 @@ static void pidLuxFloat(pidProfile_t *pidProfile, controlRateConfig_t *controlRa
...
+        errorGyroIf[axis] = constrainf(errorGyroIf[axis] + RateError * dT * pidProfile->I_f[axis] * 10, -250.0f, 250.0f);
 
         // limit maximum integrator value to prevent WindUp - accumulating extreme values when system is saturated.
         // I coefficient (I8) moved before integration to make limiting independent from PID settings
         ITerm = errorGyroIf[axis];
 
         //-----calculate D-term
-        delta = gyroRate - lastGyroRate[axis];  // 16 bits is ok here, the dif between 2 consecutive gyro reads is limited to 800
-        lastGyroRate[axis] = gyroRate;
+        delta = constrainf(RateError - lastError[axis], -800.0f, 800.0f);  // the dif between 2 consecutive gyro reads is limited to 800
```
Comment (ledvinap) [inline src/main/flight/pid.c]: This comment was probably created when difference was limited by gyro output scale.

Limiting D-term may be good idea, but this change is a bit strange ..

**Comment by borisbstyle** [comment]
I just used it from the comment. I didn't know the exact reason why it was there. The same comment is in pidrewrite as well. I guess somebody worked out luxfloat from pidrewrite, but did this part wrong.
If there is no reason for limit or another limit feel free to suggest.

**Comment by atomiclama** [comment]
May have been covered else where, not sure but.
Now that the Dterm is filtered why still have a moving average on it? I would have thought that it is now not required and will just add a delay.

**Comment by borisbstyle** [comment]
@atomiclama 
Long story short. I tested removing moving average before with not so good results and kept it in. The delay added in using gyroError is much more actually (previous situation).
I do think that possible cause of my problems were caused by derivative kick. I might retest it once more on luxfloat when filter enabled, but not sure if it is a part of this PR. It would more be something for later when stuff is merged as that is part of filtering and this is just luxfloat optimisation.

**Comment by atomiclama** [comment]
@borisbstyle 
Cheers for the explanation. Agreed it might be more optimisation at a later date.

**Comment by Spirre** [comment]
Is 50 for d on yaw really correct as default, seems high?

**Comment by borisbstyle** [comment]
@Spirre You might be right.
My intention was actually to remove the dangerous parts like high P. I might lower D for yaw as well.

**Comment by Spirre** [comment]
@borisbstyle otherwise

**Comment by borisbstyle** [comment]
Also the configurator should be modded so more decimals can be modified in luxfloat

**Comment by ethomas997** [comment]
Did a few flights with this today on [my small tricopter] (http://www.etheli.com/MiniFTC); looks really good.  Using Luxfloat, no changes to tuning, flying in Horizon mode, great control, wobbles during descents basically disappeared.  (I still need to do more tuning with the new options, etc.)  Another win for Boris B!

--ET

**Comment by Spirre** [comment]
Did some rough backyard tuning and ended up with much lower PIDs than I had when I ran PID2 on this quad before, that surprised me a bit. Had a Naze32 before and now it's a SPF3 on it.
And it seems a can lower it even more on roll.
Had hoped that I could do some real flying but now it started to rain :/
set looptime = 800

set p_pitchf =  0.850
set i_pitchf =  0.300
set d_pitchf =  0.020
set p_rollf =  0.450
set i_rollf =  0.300
set d_rollf =  0.020
set p_yawf =  3.000
set i_yawf =  0.500
set d_yawf =  0.015

set rc_rate = 110
set rc_expo = 65
set rc_yaw_expo = 10
set thr_mid = 50
set thr_expo = 0
set roll_rate = 35
set pitch_rate = 35
set yaw_rate = 50
set tpa_rate = 40
set tpa_breakpoint = 1500

**Comment by icdark** [comment]
I just tested your changes without changing PIDs. It is more responsive and bounce back on hard flips is gone. I don't have to say it is awesome :-)
:+1:

**Comment by ctzsnooze** [comment]
The default PIDs look fine to me, personally I use a bit higher on pitch/roll I but 0.3 is fine.

Wouldn't you prefer rc_yaw_expo to match rc_expo?  I've found it very nice with them both at 50, for example...  :-)

**Comment by Spirre** [comment]
@ctzsnooze Haven't seen that. I can't remember that I ever changed rc_yaw_expo, must have been an mistake :)

**Comment by Spirre** [comment]
I just flew it and I is too low, a lot of drifting. Otherwise it flew great. Now this is a very noise frame (motors are worn) and filters are heavily used. Crashed and broke an ESC, prop shaved of some components and also cut one of the SPF3 JST connectors in half  :(

**Comment by joshuabardwell** [comment]
>Wouldn't you prefer rc_yaw_expo to match rc_expo? I've found it very nice with them both at 50, for example... :-)

Personally, I want much less expo on my yaw stick than on pitch/roll. I don't like having to manage the throttle with the left stick deflected very far. I fly with RC expo at approximately 0.6, but yaw expo at about 0.1 to 0.15.

**Comment by ctzsnooze** [comment]
Of course, I agree, we want yaw and roll nicely balanced so that equal stick movements give nice 'sychronised' turns.  So when I first 'discovered' yaw expo, I went from 0 to 0.5, to match my roll expo.  Of course immediately I found less sensitivity in yaw for small stick movements - the solution was simple, just increase yaw rate to compensate for the yaw expo, until the yaw rate and roll rates 'balance nicely' with small stick movements around centre.  Once you have them balanced at small stick inputs around center again, then with equal expo you find they are still equally nicely balanced with big stick inputs.  For me this was absolutely the most amazing transformation in flying.  No matter how radical, the input, the response was entirely predictable!  Give it a try - makes a huge improvement I think.

Of course as defaults yaw is difficult to make sensible because different quads have markedly different yaw 'strength' compared to roll 'strength', so defaulting values for yaw P/rate to match roll P/rate isn't so easy.

**Comment by joshuabardwell** [comment]
>Once you have them balanced at small stick inputs around center again, then with equal expo you find they are still equally nicely balanced with big stick inputs.

But I don't _want_ them balanced because my quad has significantly less yaw authority than pitch/roll authority. I tune my rates by how fast I want the quad to spin on full stick deflection. The main limiting factor there for yaw is the copter's ability to spin without becoming destabilized and/or having unacceptable yaw jump at the end. For me, this is a yaw rate of 0.75, whereas I have pitch/roll rates at 1.5 (PIDC#1). I'm running 3S, 2204/2300kv, 6030 props, so I think I'm probably not alone in this.

**Comment by ctzsnooze** [comment]
OK I appreciate that approach.  Perhaps it may make the performance of those big sweeping banked turns at low altitude a bit difficult.  :-)  I mean like you are going in a horizontal circle say 50m radius, at high speed (so steep angle of the quad), very fast, and LOS and low to ground, it's really essential to have a consistent balance between yaw and roll responsiveness in relation to angle of the stick regardless of whether the turn is tight and fast or gently and long.  If they are not equally responsive, and worse if the responsiveness varies by angle of the stick, predictable control as you progressively tighten a fast turn into a funnel, for instance, is much more challenging.

The ability to make these synchronised turns smoothly, effortlessly and fast, and in a well balanced way, is so much easier when the proportion of roll to yaw of the quad is constant with a constant proportion of yaw and roll stick movement.

By a well synchronised turn I mean that certain 'balance' between yaw and roll that results in a perfect turn.  Let's say you move both sticks equally left.  If there is not enough yaw responsiveness, iit rolls left and falls leftwards without the nose coming around enough; too much yaw and it rotates but doesn't turn the corner.  When yaw and roll are well matched, the quad just turns perfectly.

If you want the turns you make to me predictably perfect, you need to be able to effortlessly re-create and maintain that balance in the turn as it tightens or loosens.  

I used to tune like you do until I properly understood the value of yaw expo.  Now I do, making insanely tight balanced turns is so much easier than before.

So I focus on two things when tuning relative roll and yaw rates.  First I set absolute roll rate and expo for a suitable amount of mid-stick roll sensitivity and a desired roll rate at full stick for flips.  Then I set yaw rate and expo so that it results in balanced turns with equal deflections of both sticks not just with small deflections around the centre but also at nearly full deflection for sudden hard turns.  

I can usually get that balancing done regardless of the type of quad.  Then when I switch from one to the other, the transition is very easy.   

It is nearly impossible to achieve that balance with equal stick deflections across the full range of stick deflections without function al yaw expo matching functional roll expo, because otherwise the two are only balanced at certain points.  By matching yaw and roll expo, insanely tight turns that are perfectly balanced are possible, as well as tight rolls and sudden yaw, across the full range of stick movements, with equal roll and yaw stick deflections, whether small or big.  

The end result is a quad that is predictable and simply great to fly.

I've discussed this with many other people and everyone who's tuned their quads like this never goes back to the 'old' way.

I have more than adequate absolute (full-stick) roll and yaw rates like this - on similar setups to yours.  I can't recall the exact settings I used (they vary quad to quad)  but I get more than enough yaw rate when I yaw alone, not that I yaw alone all that often.  Maybe I've been lucky but yaw jump instability hasn't been an issue for me.

I think if you give it a try you won't look back  :-)

**Comment by joshuabardwell** [comment]
> If they are not equally responsive, and worse if the responsiveness varies by angle of the stick, predictable control as you progressively tighten a fast turn into a funnel, for instance, is much more challenging.

I think you're basically right, in that, when I was increasing yaw rate, I had to also increase yaw expo so that my muscle memory of where the left stick should be vs. the right stick in a turn stayed the same. I don't think it's a given that it must be the same amount, though. I suspect that one's muscle memory gets used to whatever ratio one uses, and then anything different feels weird.

**Comment by borisbstyle** [comment]
Running these PID's and quite happy

set p_pitchf =  2.300
set i_pitchf =  0.550
set d_pitchf =  0.028
set p_rollf =  1.000
set i_rollf =  0.350
set d_rollf =  0.020
set p_yawf =  2.600
set i_yawf =  1.000
set d_yawf =  0.000
set yaw_p_limit = 500
set gyro_cut_hz = 80
set pterm_cut_hz = 50
set dterm_cut_hz = 15
set gyro_lpf = 256

**Comment by icdark** [comment]
@borisbstyle no TPA?

**Comment by borisbstyle** [comment]
@icdark 
Yes I added the rest of the settings

**Comment by ctzsnooze** [comment]
Nice.  On original Lux I can't get over P of about 1.5, would be very tight on 2.3.  Cool.

Yeah muscle memory is everything.  I have quite a few quads so I try to make them identical.  I found that with expo on yaw and roll the same it was quite easy to make the sticks do the same thing on each.  Probably we all do that anyway, for whatever relationship works over time.  No doubt in my mind that that particular muscle memory gets locked in to our personal preference for the setup of our next quad.

**Comment by borisbstyle** [comment]
Yaw and roll and pitch are not the same so you need to fiddle to get a good ratio to make them simillar.
I believe now my yaw is a bit faster. I think my roll and pitch rates will end up higher, but thats all personal :D

**Comment by borisbstyle** [comment]
So my conclusion is that the PID's don't change....well not much. You only get a more responsive stick feel. No bounce on flips and rolls etc.
Well that's my experience at least.

**Comment by borisbstyle** [comment]
I did tweak the defaults a bit based on my 4 quadcopters and some blackbox analyzing. I was still having a lot of P noise on yaw so lowered that and increased I

**Comment by borisbstyle** [comment]
In my hex files (linked above) I also added the test code for better comparing. You can easily switch in your profile between old lux and new lux.

set old_delta = 1 (enables old approach, disabled by default)

**Comment by jaapp** [comment]
Why was Luxfloat's D-Term based on the GyroError anyway?

That means that the D-Term is trying to get the change in gyro rate to 0 - which would only be of use if no stick input was given - in other words when the angular velocity has to stay the same. 

Basing it on the RateError is much more useful.

**Comment by borisbstyle** [comment]
@jaapp Exactly! Many other pid controllers have the same. Well there are a lot of articles about how control PID loop should be done. Many of those are indeed based on automated stuff, where stabilizing is more important than controlling itself. That's where you always assume control input is always zero.

**Comment by ctzsnooze** [comment]
Dterm is always attempting to get *the derivative* of gyro_rate to zero.  That's its intent.  It opposes P at all times.

For Dterm, is that the end result from targeting input-output error or targeting just output are identical *when the Dterm kicks are eliminated*.  

Think about how Dterm is calculated - it's not the magnitude of the error but the magnitude of the rate of change of the error.  After a step input PPM control change is complete, the input factor is fixed again, so the rate of change of the error becomes... the rate of change of the output.  Hence at all times apart from when there is a change in the control setpoint, both methods give the same result.  They are no different, apart from the spikes in dterm occurring on step control point changes.  See my previous post, or this simpler article:

https://barela.wordpress.com/2013/07/19/pid-controller-basics-eliminating-derivative-kick/

Bottom line is that the only difference between the two methods is the content of the kick.  Unlike Dterm, which should be opposite to Pterm, the kick is in phase with Pterm.  

Most controller theory that you read attempts to 'eliminate' derivative kick, on the basis that it is bad.

But for us it is the kick itself that - maybe just for luck - makes the error based approach much better in practice.  

That's why I'm 100% sure if we modified Boris's error based Lux float by removing the kick with a median filter, it would then no longer be any better than the original Lux code.

I can only imagine that when Lux was written, the intent was to eliminate derivative kick, thinking that it would be bad.  Actually, for us, derivative kick is good.

**Comment by borisbstyle** [comment]
@ctzsnooze 
I am not sure that it's the only difference. That's why implemented a test code to be able to compare in the field based on the profile.
In my logs the overall D shift from P is much more on the original approach, while in rateError method D seems much tighter to P. I was here looking in some older logs I have from luxfloat.
Therefore I do want to make sure I am comparing apples with apples so I am going to fly new approach....land and switch profile and fly the old approach.

**Comment by ledvinap** [comment]
@borisbstyle: The shift you observe may is probably OK:
With zero-referenced D-term P-term action is attenuated and delayed  - setpoint error must be large enough to overcome it's derivative. Also derivative sign is acting against change.
With setpoint-referenced D-term (and smoothed RC input) quad mechanics follow quite closely setpoint, D-term should be quite small (it helps to keep system output rate close to setpoint rate).

**Comment by ctzsnooze** [comment]
Very much looking forward to seeing the logs, ideally with identical sharp inputs in both systems.

If you had a switch programmed to do a standardised transient pulse input, that would be very repeatable.  I couldn't quite work out how best to do it on my Taranis but had thought of a curve that I could activate on a switch so that output in say roll was either 1250, 1500 or 1750, with a fairly steep transition between them.  Then I'd hover, hit the switch, and go quick right then recenter, quick left then recanter.  Might be useful to provide very standardised inputs for testing purposes.

**Comment by tejo** [comment]
Tried it today with @borisbstyle settings. Given I'm a total noob I was able to do flips in rate mode like a piece of cake. Banked turns are precise and crisp.  So far the best pid controller I tried. Looking fwd to get this pr merged.

**Comment by nsccap** [comment]
FYI, tried to flash the CC3D.bin file on a CC3D (currently running 1.9.0) but got "ERROR file to big for device" from the openpilot gcs (flashing from there as I still have the op boot loader on it).

**Comment by borisbstyle** [comment]
It just doesnt fit on CC3D anymore with the bootloader. That is known. You
better should put .hex on it.
Op 19 jul. 2015 12:37 schreef "Peter Kjellstrom" <notifications@github.com>:

**Comment by nathantsoi** [comment]
Fwiw, I just flashed the opbl bin on my cc3d (built: https://github.com/nathantsoi/cleanflight/tree/fly) and it works fine... and pre-tuning, ie. just taking the defaults, it flew amazing. (4s, looptime = 1200) @borisbstyle, awesome work!

**Comment by Zero9r** [comment]
For anyone using a CC3D I recommend abandoning OP bootloader and start flashing HEX files.  The OP BIN's will be missing all the elements as it takes up to much room.  Plus it causes issues with connection to GUI and such.  Like having to power on with battery and disconnect and reconnect every time you need to reboot.  Do yourself a favor and get a USB-UART, here http://www.banggood.com/CJMCU-CP2102-USB-To-TTLSerial-Module-UART-STC-Downloader-p-970993.html  Its only $3

plug it into your main port and all those problems will disappear.  HEX's are the way togo many developers are abandoning the CC3D BIN files because of it.  so help them out by not requesting BIN FIles.  its getting frustrating for both developers and users.

**Comment by nsccap** [comment]
@Zero9r I agree with your advice in general (if CF was the primary usage for the CC3D, this time it wasn't). Also do note that no one asked for a CC3D bin but (a too big) one was provided in this PR.

@nathantsoi How come your build fit? I was under the impression that some new additions on master prior to boris patch had tipped the scales...

**Comment by borisbstyle** [comment]
Added more precision in PID tuning
https://github.com/cleanflight/cleanflight/pull/1156

**Comment by borisbstyle** [comment]
Is there anybody who disagrees about the default level PID's? Those also can be changed

**Comment by lichtl** [comment]
@borisbstyle are the above linked hex files represent your current local branch?

**Comment by borisbstyle** [comment]
yes thats correct
Op 23 jul. 2015 16:11 schreef "lichtl" <notifications@github.com>:

**Comment by acceleroto** [comment]
I did a couple flights yesterday on this with a build downloaded on July 23. It flew well on my ZMR250/Cobra 2208-2000/6045/4S. Level, Horizon and Rate modes all flew well.

**Comment by jaapp** [comment]
I spend a couple of hours flying my quad today. I can vouch for this change. My ZMR 250 with crappy DYS1806 motors on 3S flew like a charm. I noticed that the range of acceptable P-gains was pretty large as well. One other factor that I noticed (coming from PID 1) is that the craft was did not have the tendency to go nose-up when I apply lots of throttle.

These nose-up issues have been haunting me for a while. The core if the issue is that my motors are crappy and put a lot of vibrations on the frame. After balancing the motor bells it flew 'ok', but as soon as a prop was less then perfect I'd have these nose-up issues again.

These issues are gone with this version of luxfloat. Just gone! 

Also, I used to have a very hard time getting luxfloat locked in. I had to reduce the D-gain to near nothing, and needed extreme rates to get it barely acceptable. Now it's just great from the get go.

**Comment by ethomas997** [comment]
I hope this mod gets merged in soon; I don't think I've heard anything but praise for it.

--ET

**Comment by ctzsnooze** [comment]
Agreed, it would be great for this to go into the master code asap.  I've tested it extensively via blackbox and in flight and it is now much easier to tune out wobbles.  Keeping I out of TPA works really well also, and the new defaults fly nicely on everything I've tried them on.

**Comment by borisbstyle** [comment]
@hydra Can you squeeze this one in 1.10 as well? This is really not a trivial change at all.

**Comment by ctzsnooze** [comment]
Yes please @hydra this is 100% the best thing to happen to Lux.  Agreed by many many people, lots of experience.  No question - please include in 1.10.

**Comment by fahrvergnuugen** [comment]
Is anyone having luck using this on the RMRC Dodo board? I reflashed with the hex file but the gyro graph in CFC stopped updating. I flashed back to the 1.9.0 SPRACING firmware and it works fine. 

Seems like the gyro doesn't like this version?

**Comment by hydra** [comment]
Will review and merge asap.

**Comment by hydra** [comment]
Merging.

**Comment by ctzsnooze** [comment]
Thanks!
